ddtrace/tracer.py

```diff
@@ -757,16 +756,7 @@ def _start_span(
             self._services.add(service)
 
         if not trace_id:
-            span.sampled = self._sampler.sample(span)
-            if not isinstance(self._sampler, DatadogSampler):
-                if span.sampled:
-                    # When doing client sampling in the client, keep the sample rate so that we can
-                    # scale up statistics in the next steps of the pipeline.
-                    if isinstance(self._sampler, RateSampler):
-                        span.set_metric(SAMPLE_RATE_METRIC_KEY, self._sampler.sample_rate)
-            else:
-                # We must always mark the span as sampled so it is forwarded to the agent
-                span.sampled = True
+            span.sampled = self._sampler.sample(span, allow_false=isinstance(self._sampler, RateSampler))
```

Comment:
How does that work? We essentially ignore so much sampling logic then don't we? Maybe I'm missing something. As a side note, if we really are ignoring some of the sampling logic, maybe we can return early in some sample functions to help with performance.

Reply:
Oh wait, `isinstance`` takes into account parent classes. Ok going to revaluate with that in mind.

Reply:
Ok I think I got it now, sorry for the confusion https://github.com/DataDog/dd-trace-py/pull/6812#discussion_r1316335311

Reply:
It's definitely confusing, I don't blame you. The new version is almost as confusing as the old version (why should a RateSampler allow false?) except for the fact that all of that confusing complexity is in the Sampler classes rather than leaking into the Tracer.